feat(core): add syntactic sugar for initializers

yjaaidi opened the pull request:
Add a helper provider functions to simplify the use of INITIALIZER intializer tokens
* `provideAppInitializer()` for `APP_INITIALIZER` 
* `provideEnvironementInitializer()` for `ENVIRONMENT_INITIALIZER`
* `providePlatformInitializer()` for `PLATFORM_INITIALIZER`

This removes the boilerplate code for the initializers like: 
```
{
   provide: APP_INITIALIZER,
   multi: true,
   useValue: initializerFn
}

```
in favor of: 
```ts
provideAppInitilizer(initializeFn)
```

This also includes a schematic migration.

## PR Checklist
Please check if your PR fulfills the following requirements:

- [X] The commit message follows our guidelines: https://github.com/angular/angular/blob/main/CONTRIBUTING.md#commit
- [X] Tests for the changes have been added (for bug fixes / features)
- [X] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: N/A


## What is the new behavior?


## Does this PR introduce a breaking change?

- [ ] Yes
- [X] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

lacolaco commented:
To keep consistency with `provideHttpInterceptors`, probably we should choose `provideAppInitializers(...initializers: InitializerFn[])`. How do you think about it?

yjaaidi commented:
> To keep consistency with `provideHttpInterceptors`, probably we should choose `provideAppInitializers(...initializers: InitializerFn[])`. How do you think about it?

Hey @lacolaco! I wonder if we really want to keep them consistent.
In most cases, you want to provide all your interceptors in one place with an array that makes the interceptors order explicit and easy to understand.

On the other hand, app initializers are often totally independent and even provided in different places like custom provide functions. (e.g. `provideStore() { return [..., provideAppInitializer(store => store.init())] }`
My guess is that in most cases there will be only one app initializer so writing `provideAppInitializers([() => {}])` can be syntactically heavy with too many types of brackets in one line 😊

We could end up in a `styleUrls / styles` situation where the array was of a cardinality of one in most cases.

lacolaco commented:
@yjaaidi As you says, each APP_INITIALIZER function is basically independent. I agree on your design 👍

JeanMeche commented on packages/core/src/application/application_init.ts:
```diff
@@ -135,6 +135,38 @@ export const APP_INITIALIZER =
+    useFactory: () => {
+      const injector = inject(Injector);
+      return () => {
+        return runInInjectionContext(injector, initializerFn);
```
That would be even better !

alxhub replied:
👍 either way works for me.

atscott commented:
@yjaaidi can you please update the commit to be a `feat` rather than `refactor`?

alxhub reviewed (approved):
Reviewed-for: public-api

yjaaidi commented:
> 👍 either way works for me.

@alxhub done here as there is a low risk this breaks anything and rolling back to the previous implementation is straightforward.

Now, all app initializers will run with injection context.

@atscott do you want me to squash the commits or keep them separated?

atscott reviewed (approved):
reviewed-for: public-api

yjaaidi commented:
Thanks @atscott, @alxhub and @JeanMeche for your time!

JoostK reviewed (commented):
Feel free to address in a follow-up as these are non-blocking.

yjaaidi commented on packages/core/src/application/application_init.ts:
```diff
@@ -135,6 +135,33 @@ export const APP_INITIALIZER =
+ *         myInitializer();
+ *      }
+ *   }
+ * }
```
✅ Done

yjaaidi commented on packages/core/src/application/application_init.ts:
```diff
@@ -175,7 +203,7 @@ export class ApplicationInitStatus {
     const asyncInitPromises = [];
     for (const appInits of this.appInits) {
-      const initResult = appInits();
+      const initResult = runInInjectionContext(this.injector, () => appInits());
```
✅ Done

When it comes to public APIs, I prefer paying the cost of closure to avoid implicit args forwarding. What do you think?
I changed it anyway 😊

JoostK replied:
That is a valid concern. In this case I think it's fine not to have the closure

yjaaidi commented:
Now that the initializer is running within an injection context even without using the helper, I update the reference docs with simplified examples. What do you think?

markgoho commented on packages/core/src/application/application_init.ts:
````diff
@@ -135,6 +141,31 @@ export const APP_INITIALIZER =
+ * }
+ * ```
+ *
+ * Note that he provided initializer is run in the injection context.
````
```suggestion
 * Note that the provided initializer is run in the injection context.
```

AndrewKushnir commented on packages/core/src/application/application_init.ts:
```diff
@@ -135,6 +141,31 @@ export const APP_INITIALIZER =
+ */
+export function provideAppInitializer(
+    initializerFn: () => Observable<unknown>| Promise<unknown>| void): Provider {
+  return {
```
I think we should update the function to return `EnvironmentProviders`, similar to this methid:

https://github.com/angular/angular/blob/a3787bdfa61ef98c9e109d62c49e7bf873c03807/packages/platform-server/src/provide_server.ts#L29-L34

The goal of this change would be to have an improved type safety, which would prevent accidental use of this function in `@Component.providers` (where it would have no effect anyways).

With this change, we can also consider deprecating `APP_INITIALIZER` token in the future (**not** in this PR) in favor of this function for the same reasons: better type safety and avoiding a situation when `multi: true` is accidentally forgotten.

@alxhub @atscott WDYT?

atscott replied:
Good call. 👍

yjaaidi replied:
Good one! Honestly, didn't think of it at all!

✅ Done!

Do you have a specific convention for type tests or the one I added is fine?

JeanMeche commented:
As discussed with the team and Younes, we would want to extend this behavior to other tokens :

* `ENVIRONMENT_INITIALIZER` => `provideEnvironementInitializer()`
*  `PLATFORM_INITIALIZER` => `providePlatformInitializer()`

This way we would have a consistent API around those tokens and we could deprecate them (in order to remove them from the public API) leaving only the provider functions. 

Alongside this deprecation, we would want to have a migration schematic to smooth out the transition and give a clear direction to the developer.

yjaaidi commented:
Hey @JeanMeche, I was thinking about the migrations scenarios for the schematic, and I wonder what would we should do for the following extreme cases:

### A. no multi (really rare case but happens in some Angular core tests for example :wink:)
```ts
{provide: APP_INITIALIZER, useValue: [initializer1, initializer2]}
```

1. Should we just do nothing?
2. Should we add a "todo" comment?

### B. logic in the factory
```ts
{
  provide: APP_INITIALIZER,
  useFactory() {
    const a = inject(A);
    const b = inject(B);
    const value = b.doSomething();
    return () => {
      return a.initialize(value);
    };
  },
  multi: true
}
```

1. Should we move everything from the factory inside the initializer?
2. Should we move stuff from the factory inside the initializer only if there are only calls to `inject()` and skip the transform otherwise?
In case the transform is skipped, should we add a "todo" comment?

yjaaidi commented:
There is also the case of a more dynamic factory:

```ts
{
  provide: APP_INITIALIZER,
  multi: true,
  deps: [Something],
  useFactory(something: Something) {
    return something ? initializerA : initializerB;
  }
}
```

Maybe the best strategy is to try to transform the most simple cases (`useFactory` + inline function) like this:
```ts
{
  provide: APP_INITIALIZER,
  multi: true,
  deps: [A],
  useFactory(a: A) {
    return () => a.init();
  }
}

// becomes:
provideAppInitializer(() => {
  const a = inject(A);
  return a.init();
})
```

Otherwise, we use the factory as an IIFE (except that we replace `deps` with `inject` calls):
```ts
{
  provide: APP_INITIALIZER,
  multi: true,
  deps: [Something],
  useFactory(something: Something) {
    return something ? initializerA : initializerB;
  }
}

// becomes:
provideAppInitializer((() => {
  const something = inject(Something);
  return something ? initializerA : initializerB;
})());
```

yjaaidi commented:
Done!
- ✨ add `provideAppInitializer`
- ✨ add `provideEnvironmentInitializer`
- ✨ add `providePlatformInitializer`
- ✨ add migration schematic

Pending questions:
- should we deprecate `APP_INITIALIZER`, `ENVIRONMENT_INITIALIZER`, `PLATFORM_INITIALIZER`?
- which Angular version should be the migration target? Meanwhile, I put `18.0.0`.
- in the extreme case where `useFactory` is used and has some logic executed in the factory, is it ok to accept the risk of moving that logic into the initializer (because we are just moving the whole `useFactory` function in an IIFE within the initializer) or should we skip the migration and add a `TODO` comment?
Here is a random example of the usage mentioned above: https://github.com/PillowPillow/ng2-webstorage/blob/504296cb0cfaede59996bbc3e2f9706bea77f54a/projects/ngx-webstorage/src/lib/module.ts#L13

JeanMeche commented:
* I believe yes the idea was to deprecate (as to not have 2 ways to achieve the same feature). 
* v18 is right for the migration
* I'll let @AndrewKushnir or @alxhub answer on this.

JeanMeche commented on packages/core/src/application/application_init.ts:
```diff
@@ -135,6 +141,32 @@ export const APP_INITIALIZER =
         ngDevMode ? 'Application Initializer' : '');
 
+/**
+ * @publicApi
```
nit: usually we have the `publicApi` attribute at the bottom.

yjaaidi replied:
Done!

JeanMeche commented on packages/core/src/application/application_init.ts:
```diff
@@ -135,6 +141,32 @@ export const APP_INITIALIZER =
+/**
+ * @publicApi
+ *
+ * This is syntactic sugar for the following:
```
We need this to be `@usageNote` to appear on angular.dev.

yjaaidi replied:
Done!

JeanMeche commented on packages/core/src/application/application_init.ts:
````diff
@@ -135,6 +141,32 @@ export const APP_INITIALIZER =
+ *   useValue: initializerFn
+ * }
+ * ```
+ *
````
we also need a short `@description` so the docs doesn't feel so empty. 
See : https://ng-dev-previews-fw--pr-angular-angular-53152-adev-prev-mms8d1sk.web.app/api/core/provideAppInitializer

yjaaidi replied:
Done!

JeanMeche commented on packages/core/src/di/provider_collection.ts:
```diff
@@ -35,6 +35,30 @@ export function makeEnvironmentProviders(providers: (Provider|EnvironmentProvide
 }
 
+/**
+ * @publicApi
```
Same a `provideAppInitializer`

* move `@publicAPI` to the bottom of the docs
* Make `@description` for the adev docs
* Nit: provide a short `@usageNotes`

yjaaidi replied:
Done, but not really happy with the example. Maybe an example with the router would be more interesting.

JeanMeche commented on packages/core/src/platform/platform.ts:
```diff
@@ -150,7 +158,35 @@ export function createOrReusePlatformInjector(providers: StaticProvider[] = []):
+ *
+ * Note that the provided initializer is run in the injection context.
+ *
+ * @see {@link PLATFORM_INITIALIZER}
```
Same a `provideAppInitializer`

* move `@publicAPI` to the bottom of the docs
* Make the use of `@description` for the adev docs
* Nit: provide a short `@usageNotes`

yjaaidi replied:
Done!

yjaaidi commented:
> I believe yes the idea was to deprecate (as to not have 2 ways to achieve the same feature).

Perfect! Is there a convention for indicating the sunset version? or do you want to introduce one? 😇

E.g.
```ts
/**
 * @deprecated from v18.0.0 ...
 * @sunset v19.0.0
 */
```

or to avoid non-jsdoc block tags:
```ts
/**
 * @deprecated from v18.0.0 (sunset v19.0.0)
 */
```

The sunset could be postponed but it should encourage the users to migrate more than a deprecation and also it helps to quickly find what should be dropped during the release process of a new major.

This could help maintain this page: https://angular.io/guide/deprecations#index
BTW, I can't find it on https://angular.dev, is it a todo or did I miss it?

JeanMeche commented:
The deprecation page has been dropped from ADEV for maintenance reasons. 
Also there is no announced sunset version. The only policy is to support at least the deprecated entry for 2 majors. Deprecation in v18 can only be removed from v20 onwards.

yjaaidi commented:
Ok. Thanks for the policy reminder 😉
That said, I still think that it would be nice to indicate the sunset (which is the "maybe removed in" column) beyond the policy.
For instance, some deprecations are hard to get rid of and it would be nice to reassure people that the sunset is at least for 3 or 4 majors.

Is there a convention to indicate the version of deprecation (e.g. from v18.0.0)?

JeanMeche commented:
AFAIK we don't have any convention but we have existing examples across the code base. 

https://github.com/angular/angular/blob/571a61beeb0edab23eccc172193c5c44ab4aa339/packages/forms/src/directives/reactive_directives/form_control_directive.ts#L80-L81

https://github.com/angular/angular/blob/571a61beeb0edab23eccc172193c5c44ab4aa339/packages/core/testing/src/test_bed.ts#L112-L113

https://github.com/angular/angular/blob/571a61beeb0edab23eccc172193c5c44ab4aa339/packages/core/src/di/r3_injector.ts#L108-L112

You could do something similar. Also the deprecation message should indicate what the new alternative/recommended way is.

yjaaidi commented:
Yeah, that's what I observed and why I thought it would be nice to establish a convention on the deprecation version.

yjaaidi commented:
I updated the jsdoc and added the deprecations.
That said, I don't understand why the tests are failing here https://github.com/angular/angular/actions/runs/8726611240/job/23942284619?pr=53152

They pass locally as I updated the public API golden after adding the deprecations. Could that be the cache? Or did I miss something?
What do you think @JeanMeche?

JeanMeche commented:
You need to update the golden files with `yarn bazel run //packages/core:core_api.accept` !

yjaaidi commented:
> You need to update the golden files with `yarn bazel run //packages/core:core_api.accept` !

I did and you can see that the golden are updated properly (i.e. `@deprecated` added): https://github.com/angular/angular/pull/53152/files#diff-3395003fb035cfc47bf17a421174501f2bf7a3ed164cd3d0fbc4d91f50475f2b
and the test is passing locally.

That's why I don't get it.

JeanMeche commented:
@yjaaidi You'll need to rebase on main and rerun the script. You'll then see the delta.

yjaaidi committed:
feat(core): add syntactic sugar for initializers

add helper functions provideAppInitializer, provideEnvironmentInitializer & providePlatformInitializer
to respectively simplify and replace the use of APP_INITIALIZER, ENVIRONMENT_INITIALIZER, PLATFORM_INITIALIZER

add a migration for the three initialiers

yjaaidi commented:
@JeanMeche I don't understand how rebasing would have fixed it. AFAIK the golden check just makes sure that the goldens are up-to-date, right? 
I rebased and doesn't seem to fix it 🫤. (note that the test is passing locally)

BTW, I also noticed that `APP_INITIALIZER` and friends disappeared from angular.dev. Does angular.dev remove deprecated symbols from the reference?

JeanMeche commented:
On ADEV, the deprecated filter is off by default, you have to toggle it do display deprecated APIs.

yjaaidi commented:
Apparently, I didn't rerun the goden generation after the rebase. (the rebase didn't put the `provideExperimentalZonelessCD` in the right alphabetical order 😊)
This time it should work.

That being said, I still don't understand why it failed before the rebase.

SerkanSipahi commented:
@yjaaidi @JeanMeche 👀

any updates here?

tomalaforge commented:
any updates on this helper function?

yjaaidi commented:
> @yjaaidi any updates on this helper function?

Hey buddy! What a timing.
I just updated the PR between two flights 😅 and I hope that this will ship in v19.

JeanMeche reviewed (approved):
From a docs perspective.

alxhub commented on packages/core/schematics/migrations.json:
```diff
@@ -9,6 +9,11 @@
       "description": "Updates ExperimentalPendingTasks to PendingTasks",
       "factory": "./bundles/pending-tasks#migrate"
+    },
+    "provide-initializer": {
```
I don't think this migration is safe to run automatically. The provider functions in this PR return `EnvironmentProviders` (as they should) but existing usages may not be typed in a way to handle that type.

We can exclude it and use it as an optional migration for now.

yjaaidi replied:
What would be the solution to make it automatic for v20?
We could detect where it is used and transform when it is safe. When it's not safe what would be the fallback? A type cast? A compact function? 🤔

yjaaidi replied:
Thx @JeanMeche for handling this while I'm AFK 😊

JeanMeche committed:
fixup! feat(core): add syntactic sugar for initializers

AndrewKushnir commented:
[Presubmit](http://test/OCL:688544720:BASE:688544746:1729608036448:61db412).

AndrewKushnir commented:
This PR was merged into the repository by commit 19edf2c057f7587bc16812685d31a556521ad414.

The changes were merged into the following branches: main

pawelfras commented on packages/core/src/application/application_init.ts:
```diff
@@ -25,7 +34,12 @@ import {isPromise, isSubscribable} from '../util/lang';
  *
+ * Note that the provided initializer is run in the injection context.
+ *
+ * @deprecated from v18.1.0, use provideAppInitializer instead
```
```suggestion
* @deprecated from v19.0.0, use provideAppInitializer instead
```

pawelfras commented on packages/core/src/application/application_tokens.ts:
```diff
@@ -49,6 +49,11 @@ const DEFAULT_APP_ID = 'ng';
 /**
  * A function that is executed when a platform is initialized.
+ *
+ * @deprecated from v18.1.0, use providePlatformInitializer instead
```
```suggestion
* @deprecated from v18.1.0, use providePlatformInitializer instead
```

pawelfras reviewed (commented):
Hi,
thank you @yjaaidi for your great work!
The deprecation version mentioned in JSDocs for `APP_INITIALIZER` and `PLATFORM_INITIALIZER` is 18.1.0. Shouldn't this be changed to 19.0.0 as it was done for `ENVIRONMENT_INITIALIZER`?
If so, here's my PR with this small adjustment: https://github.com/angular/angular/pull/58322

yjaaidi commented:
Nice catch Pawel!

Yogu commented on packages/core/src/di/provider_collection.ts:
````diff
@@ -50,6 +49,43 @@ export function makeEnvironmentProviders(
+ * ```
+ * createEnvironmentInjector(
+ *   [
+ *     provideEnvironmentInjector(() => {
````
Should this be `provideEnvironmentInitializer`?

yjaaidi replied:
Oh good one! Would you like to make a PR to fix this?

davidlj95 replied:
Just opened https://github.com/angular/angular/pull/58355 :)
